### PR TITLE
OpenTelemetry tracing, with jaeger in development environment.

### PR DESCRIPTION
**Pull request by mike-kingsbury**
For developers - includes an instance of Jaeger in the docker-compose development environment, available at port 16686.

For production - supports OTLP/http exported traces to a specified endpoint in the configuration files.

A combination of auto-instrumentation and manual instrumentation in the worker where external actions are occurring. Traces are sent to OTel collectors based on an environment variable, or in the case of development environment, a jaeger instance brought up with docker-compose.

Please see docs/tracing.md for more details.

# Maintainers will complete the following section

- [x] Commit messages are descriptive enough
- [x] Code coverage from testing does not decrease and new code is covered
- [ ] New code has type annotations
- [ ] ~OpenAPI schema is updated (if applicable)~
- [ ] ~DB schema change has corresponding DB migration (if applicable)~
- [x] README updated (if worker configuration changed, or if applicable)
- [ ] Draft release notes are updated before merging

### CODE REVIEW

File: cachito/web/app.py [hunk 1]
```diff
@@ -133,10 +144,62 @@ def create_app(config_obj=None):
...
+    _instrument_app(app)
     return app
 
 
+def _instrument_app(app):
+    """
+    Instrument the Flask app.
+
+    Returns:
+    A flask app object
```
Comment (brunoapimentel) [inline cachito/web/app.py:156]: In fact, this function is not returning anything, so I'd drop this part here.

File: docs/tracing.md [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+
+
+
+[1]: https://opentelemetry.io/docs/concepts/signals/traces/
+[2]: https://pypi.org/project/opentelemetry-instrumentation-requests/
+[3]: https://pypi.org/project/opentelemetry-instrumentation-flask/
+[4]: https://pypi.org/project/opentelemetry-instrumentation-sqlalchemy/
+[5]: https://pypi.org/project/opentelemetry-instrumentation-celery/
+[6]: https://localhost:16686/
+
```
Comment (brunoapimentel) [inline docs/tracing.md:23]: This file has two trailing whitelines.

File: requirements-test.in [hunk 3]
```diff
@@ -2,4 +2,12 @@ jsonschema
...
 pytest-cov
+pytest-opentelemetry
 pyyaml
+opentelemetry-sdk
+opentelemetry-instrumentation-celery
+opentelemetry-instrumentation-requests
+opentelemetry-instrumentation-sqlalchemy
+opentelemetry-exporter-jaeger
+opentelemetry-exporter-otlp-proto-http
+
```
Comment (brunoapimentel) [inline requirements-test.in:13]: Same as above: two trailing whitelines.

File: tox.ini [hunk 4]
```diff
@@ -86,7 +86,12 @@ basepython = python3
...
+    pytest -rA -vvvv --export-traces \
         --confcutdir=tests/integration \
         {posargs:tests/integration}
-passenv = KRB5CCNAME, REQUESTS_CA_BUNDLE, KRB5_CLIENT_KTNAME, CACHITO_TEST_CERT, CACHITO_TEST_KEY, JOB_NAME
+passenv = KRB5CCNAME,REQUESTS_CA_BUNDLE,KRB5_CLIENT_KTNAME,CACHITO_TEST_CERT\
+          CACHITO_TEST_KEY,JOB_NAME
+setenv =
+    OTEL_RESOURCE_ATTRIBUTES=service.name=cachito-integration-tests
+    OTEL_EXPORTER_OTLP_ENDPOINT=http://localhost:4317
+
```
Comment (brunoapimentel) [inline tox.ini:97]: Same as above: two trailing whitelines.

File: docs/tracing.md [hunk 5]
```diff
@@ -0,0 +1,23 @@
+# Tracing
```
Comment (brunoapimentel) [inline docs/tracing.md:1]: It would be nice to include a link to this doc in https://github.com/containerbuildsystem/cachito#more-documentation.

Comment (mike-kingsbury) [reply docs/tracing.md:1]: Added.

File: requirements-test.in (hunk 3 above)
Comment (mike-kingsbury) [reply requirements-test.in]: Cleaned up.

File: docs/tracing.md (hunk 2 above)
Comment (mike-kingsbury) [reply docs/tracing.md]: Cleaned up.

File: tox.ini (hunk 4 above)
Comment (mike-kingsbury) [reply tox.ini]: Cleaned up.

**Comment by brunoapimentel** [comment]
I also noticed that coverage on [cachito/workers/tasks/celery.py](https://coveralls.io/builds/63472985/source?filename=cachito%2Fworkers%2Ftasks%2Fcelery.py#L25) fell significantly. Can we make a simple test to cover that file?

**Comment by mike-kingsbury** [comment]
> I also noticed that coverage on [cachito/workers/tasks/celery.py](https://coveralls.io/builds/63472985/source?filename=cachito%2Fworkers%2Ftasks%2Fcelery.py#L25) fell significantly. Can we make a simple test to cover that file?

Pragma no-cover'ed the function; its all calls to the OpenTelemetry framework and has no good way to test or observe the effects in a unit test.  (And I will press the comment button, and not the confusing 'Close with comment' button that I have mistaken for 'close the thread'.!)

File: docker-compose.yml [hunk 6]
```diff
@@ -114,6 +114,16 @@ services:
       - nexus
       - rabbitmq
 
+  jaeger:
+    image: jaegertracing/all-in-one:latest
```
Comment (lkolacek) [inline docker-compose.yml:118]: Do we want to specify `docker.io` registry for the image here as in other cases in `docker-compose.yml` so during the run we are not prompt  to pick one?
```
? Please select an image: 
    registry.fedoraproject.org/jaegertracing/all-in-one:latest
    registry.access.redhat.com/jaegertracing/all-in-one:latest
  ▸ docker.io/jaegertracing/all-in-one:latest
    quay.io/jaegertracing/all-in-one:latest
```

Comment (mike-kingsbury) [reply docker-compose.yml]: Good idea! Updated.

**Review by brunoapimentel — APPROVED** [review]
LGTM :rocket::rocket::rocket:

**Review by lkolacek — APPROVED** [review]
LGTM thank you Mike! :+1:

**Pull request merged**